#**Tarea 1: Repaso de los conceptos básicos de Python**
## ***Ciencia de datos para físicos: Facultad de ciencias, UNAM***
###*Alumno: Jesús Abraham Bautista López*

**Ejercicio 1 (1 punto):**

 * ¿Qué es una palabra reservada en python? Enlista 5 ejemplos de nombres reservados en python y describe su función.

* Coloca tres ejemplos de nombres de variables que no son válidos en python. Explica por qué no son válidos.

Una palabra reservada son aquellas palabras que como su nombre indica ya estan 
reservadas en el lenguaje de programación, en este caso Python, para hacer tareas especificas.

 Como ejemplos podemos tener:
* **lambda**: que es una variable matematica 
* **True**: que es un valor booleano, y es el resultado de una operacion de comparación, True = 1 en booleano.
* **False**: el mismo caso que en la lina anterior, False equivale a 0 en booleano.
* **class**: es una abstraccion de un objeto en el mundo real, proveen una fora de empaquetar datos.
* **while**: que es una forma de ejecutar la misma tarea varias veces dependiendo de como definamos una condición.

Variables no válidas en python: 
 
 * **"5perro = 10"** (NO es valida ya que inicia con un número).
 * **"nu mero = 99"** (NO es válida ya que tiene un espacio).
 * **"mi-edad = 10.3"** (NO es válida por que tiene un guion, solo son válidos los guiones bajos).

**Ejercicio 2 (1 punto):**

**En una celda de código, escribe una función que encuentre las raíces de cualquier polinomio de segundo grado (implementar la chicharronera). En una celda markdown detalla posibles casos en la cual la función podría fallar. Usa dicha función para encontrar las raíces del polinomio $f(x) = 10x^2 - 2x$
 e imprimelas en pantalla.**

In [ ]:
#Recordamos que para calcular la "chicharronera" nececitamos los valores de 3 variables, "a" para el coeficiente cuadrado,
#"b" para el coenficiente lineal y "c" término independiente.
a= float(input('Introduzca el valor de a: '))
b= float(input('Introduzca el valor de b: '))
c= float(input('Introduzca el valor de c: '))

#Ahora se va a programar la formula, recordando que tendremos dos valores (x1 y x2) diferentes debido al termino +/-
x1=(-b+(((b**2)-(4*a*c))**0.5))/(2*a)
x2=(-b-(((b**2)-(4*a*c))**0.5))/(2*a)

#Una vez calculados ambos valores, procedemos a imprimirlos en la pantalla
print('El valor de x1 es: ')
print(x1)

print('El valor de x2 es: ')
print(x2)

Introduzca el valor de a: 10
Introduzca el valor de b: -2
Introduzca el valor de c: 0
El valor de x1 es: 
0.2
El valor de x2 es: 
0.0


La funcion podría fallar si introducimos una letra en alguno de los  valores a,b o c. Pues antes declaramos que estas serían flotantes, numeros decimales o enteros. También si introducimos fraciones $(2/2)$, de la misma manera si a la variable "a" le asignamos un *cero*.

**Ejercicio 3 (1 punto):**

*Caida libre*: **El gran Galileo Galilei subió la torre de Pisa para determinar el tiempo que tarda un objeto en caer desde una altura $H$ al suelo. Para ello, colocó un objeto de masa $m=100$ kg en la torre y lo dejó caer. El tiempo que tardó en caer fue $t=0.05616667$ minutos.
Considerando que la aceleración de la gravedad es $g=9.8$ $\frac{m}{s^2}$, escribe un programa que determine la altura de la torre de Pisa. El programa debe imprimir la altura en la pantalla.
Recuerda que el movimiento de caída libre simplificado cumple la ecuación: 
$y(t) = H + vt - \frac{1}{2}gt^2$.**

In [ ]:
#Para este programa tendremos presente la ecuación y(t), donde vemos que "v" es la velocidad inicial del objeto, para nuestro caso este valor es cero
#pues al ser caida libre, cae desde el reposo  (se deja caer), entonces el valor "vt"=0. También debemos tomar en cuenta qué
#cuando el objeto llegue al suelo y = 0 m. Ahora con esas consideraciones, procedemos a pedir el tiempo en el cae el objeto "tm":

tm = float(input('Introduzca el tiempo que tardó en caer el objeto (en minutos): '))
#Una vez tenemos el valor en minutos "tm", se convierte a  segundos "ts":
g= 9.8
y= 0
ts= (tm*60)/(1)
#Por último evaluamos H con los valores conocidos e imprimimos el valor de la altura de la torre de ´Pisa
H= y + ((g)*(ts**2))/(2)
print('El valor de la altura (H) de la que cayó el objeto es: ')
print(H)

Introduzca el tiempo que tardó en caer el objeto (en minutos): 0.05616667
El valor de la altura (H) de la que cayó el objeto es: 
55.64881660520021


**Ejercicio 4 (1 punto):**

Crea una lista de números enteros del 0 al 99 : 
`lista = [0,1, 2, 3, 4, ... , 95, 96, 97, 98, 99]` (6 puntos menos si creas la lista manualmente).

Escribe un programa que imprima en la pantalla una la lista que cumpla con las siguientes condiciones:
* Los ultimos 10 elementos: debe imprimir lo siguiente `[90, 91, 92, 93, 94, 95, 96, 97, 98, 99]`.
* Los primeros 11 elementos: debe imprimir lo siguiente `[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]`.
* La serie de elementos de la lista que están entre 60 y 75: debe imprimir lo siguiente `[60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75]`.
* El número 50: debe imprimir lo siguiente `50`.
"Debes aplicar el concepto de *slicing* para resolver este ejercicio (notación de puntos `[:]`).
"Puntos menos si se crean listas nuevas para cada caso y/o se seleccionan manualmente los valores.

In [ ]:
lista = list(range(0,100)) #con esta línea creamos una lista que va del 0 al 99 (un total de 100 valores)

print(lista[90:]) #esta línea imprimirá los valores a partir del 90 de nuestra anterior lista

print(lista[:11]) #esta línea imprimirá los valores a partir del 0 al 10 de nuestra lista

print(lista[60:76]) #esta línea imprimirá los valores del 60 al 75 de nuestra lista

lista[50] #esta línea imprimirá el valor "50" que se encuentra en nuestra lista

[90, 91, 92, 93, 94, 95, 96, 97, 98, 99]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75]


50

**Ejercicio 5 (1 punto):**

Crea una lista de números enteros del 0 al 1000. A partir de dicha lista:

* Guarda en una variable una lista con los números pares usando ciclos y condicionales. Imprime los últimos 10 elementos de la lista final.

* Guarda en una variable una lista con los números impares usando listas comprensivas (comprehension list). Imprime los primeros 10 elementos de la lista final.

In [ ]:
#Primera parte: ciclos y condicionales
lista_del_0_al_1000 = list(range(0,1001)) #con esta línea creamos una lista que va del 0 al 1000 (un total de 100 valores)

print('Lista del 0 al 1000', lista)
def extraer_pares(lista): #definimos esta funcion que nos servirá para tomar los número pares de una lista
  pares = [] #creamos la variable vacia donde se guardaran los números pares de nuestra primer lista
  for n in lista: #lo siguiente es establecer un ciclo para obtener los numero pares de nuestra lista (iteramos)
    if n %2==0:    #calculamos el módulo de nuestra n, tal que nos brinde solo los valores pares
      pares.append(n) #si se cumple lo anterior, a nuestra variable "pares" le agregamos el n actual
  return pares  #regreamos a nuestra variable ya ocupada por las n

resultado = extraer_pares(lista_del_0_al_1000[982:]) #aqui invocamos la funcion declarada anteriormente, donde el argumento es nuestra primer lista
#no olvidando que como solo queremos los últimos 10 elementos, tomaremos los valores a partir del 982 en adelante
print('Lista de los últimos 10 número pares del 0-1000: ', resultado) #por último solo imprimimos el contenido de la variable números. 
#print(len(resultado)) ESTA LINEA SOLO SE USO PARA CONTAR LOS TERMINOS DE NUESTRA LISTA, POR ENDE SE GUARDA COMO UN COMENTARIO

#Segunda parte: compresivas
lista_0_a_1000 = [numero for numero in #creamos una la lista sobre la que se hará la compresion
            [numero for numero in range(0,1001)] #declaramos el rango a seguir
                if numero % 2 != 0 ] #Ahora solo la condición para que tome los números impares
print('Lista de los primeros 10 números impares del 0 al 1000: ' ,lista_0_a_1000[:10])  #Por útimo imprimimos el rango de valores impares que queremos

Lista del 0 al 1000 [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217,

**Ejercicio 6 (2 puntos): Aproximación de $\pi$**

Una de las formas más sencillas de aproximar $\pi$ es mediante la siguiente serie infinita de Leibniz:
$$\frac{\pi}{4} = \sum_{n=0}^{\infty}\frac{(-1)^{n}}{(2n+1)} = \frac{1}{1} - \frac{1}{3} + \frac{1}{5}-\frac{1}{7}+\frac{1}{9}-...$$
crea una función para aproximar $\pi$ utilizando la serie de Leibniz. Usa n = 100000 e imprime el valor de tu aproximación tanto el valor de $\frac{\pi}{4}$ como $\pi$.
El valor de $\pi$ se obtiene al despejar:
$$\pi = 4 * \left(\sum_{n=0}^{\infty}\frac{(-1)^{n}}{(2n+1)} \right)$$"
La siguiente celda de codigo ofrece el valor de $\frac{\pi}{4}$.

In [ ]:
import numpy as np
#El valor de pi/4
print("El valor de pi/4 es aproximadamente {}".format(np.pi/4))

El valor de pi/4 es aproximadamente 0.7853981633974483


El valor de pi/4 es aproximadamente 0.7853981633974483.

In [ ]:
#Calculo de pi y pi/4 usando la serie infinita de Leibniz
def termino (n): #definimos una funcion que nos dará la formula despejada anteriormente
  resultado = ((-1)**n)/(2*n+1) #escribimos la "formula" que hará el cálculo para el término n que buscamos, teniendo cuidado con los parentesís
  #para respetar la jerarquia de operaciones

  return (resultado) #regresamos el resultado
  #Con la anterior función ya podemos obtener el valor de cualquiera de los términos de la serie de Leibniz

#Lo siguiente será hacer la "suma" de todos los términos 
suma_L = 0 #Esta variable corresponde a la serie de leibniz, iniciará desde cero
for i in range(0,100000): #con este for generaremos varios números en el rango de presicion que busquemos (n)

  suma_L = suma_L + termino (i) #Ahora definimos una nueva suma que arranca desde lo que teniamos en la suma de antes (0) más el término íesimo 

#Al final obtenemos el valor de pi/4 en esa variable, por lo que solo hacemos un despeje al declarar la variable que nombraremos pi
pi = suma_L * 4
print('El valor de pi obtenido por la serie de Leibniz es: ', pi) #Por último imprimimos el valor calculado de pi
print('El valor de pi/4 obtenido por la serie de Leibniz es:', pi/4) #Y el valor de pi/4


import numpy as np
#El valor de pi/4
print("El valor de pi/4 es aproximadamente {}".format(np.pi/4))#comparamos el valor de pi/4 de python con el obtenido en nuestró código (Nos acercamos)

El valor de pi obtenido por la serie de Leibniz es:  3.1415826535897198
El valor de pi/4 obtenido por la serie de Leibniz es: 0.7853956633974299
El valor de pi/4 es aproximadamente 0.7853981633974483


**Ejercicio 7 (1 puntos): Identificador de vocales**

Escribe una función que reciba una cadena, por ejemplo 'el pErro de la tiA chOnita' y regrese la cantidad de vocales que contiene. La función debe contener varios casos de prueba, if, elif y else. Con la cadena de ejemplo, imprime en pantalla el resultado de la función (10).

Para conocer si un caracter es una vocal, primero debes descomponer la cadena en caracteres individuales. Para ello observa el siguiente código:

In [ ]:
# Divide una cadena en caracteres individuales
cadena_prueba = 'el pErro de la tiA chOnita'
caracteres_individuales = list(cadena_prueba)
print(caracteres_individuales)

['e', 'l', ' ', 'p', 'E', 'r', 'r', 'o', ' ', 'd', 'e', ' ', 'l', 'a', ' ', 't', 'i', 'A', ' ', 'c', 'h', 'O', 'n', 'i', 't', 'a']


In [113]:
#Para este código usaremos el principio del código de arriba, pero en vez de hacer una lista de los caracteres, usamos 'lower' para convertir las vocales
#MAYUSCULAS ingresadas a minusculas, una vez hecho eso, creamos un bucle for en una funcion 'contador' que evalúa una condición y luego ejecuta un bloque 
#de código si la condición 'contador' es verdadera. El bloque de código se ejecuta repetidamente hasta que la condición llega ser o es falsa.

def contar_vocales(cadena): #En caso de que la letra esté dentro de 'aeiou', se aumenta el contador. Al terminar la función regresamos el contador.
#Así es como contamos las vocales de una cadena en Python
	contador = 0
	for letra in cadena:
		if letra.lower() in 'aeiou': 
			contador += 1 #esta función devolverla un entero y recibirá una cadena
	return contador #hacemos un retorno a nuestra funcion

cadena = input('Ingrese una cadena, ya sea en mayusculas o minusculas: ') #se le pide al usuario que ingrese una frase (cadena)
cantidad = contar_vocales(cadena) #en esta variable solo guardaremos el valor de la funcion antes programada
print(f'En la cadena ', cadena, 'hay', cantidad, 'vocales') #por último solo imprimimos tanto la cadena como el número de vocales de la misma

Ingrese una cadena, ya sea en mayusculas o minusculas: el pErro de la tiA chOnita
En la cadena  el pErro de la tiA chOnita hay 10 vocales


**Ejercicio 8 (2 puntos):**

Escribe una clase "calculadora" que regrese el resultado de las operaciones de suma, resta, multiplicacion, division, modulo y potencia. La clase debe contener un método para cada operación.

Los métodos que pudiesen presentar errores como division entre cero, deben ser manejados de tal manera que el programa no imprima ningun error (vimos una clausula para hacer esto).

Adicional a esto, los métodos modulo y potencia deben ser generales, es decir, deben funcionar para cualquier número entero positivo (es un parámetro extra del método). Por ejemplo, la potencia de 2 elevado a 3 es 8, la potencia de 5 elevado a 3 es 125, etc.

Crea una instancia de la clase calculadora y prueba cada uno de los métodos, con valores de tu elección. Para los métodos que pudiesen presentar errores da dos ejemplos, uno totalmente funcional (3/5 por ejemplo) y otro que sea conflictivo pero que no regresa ningún error (4/0 por ejemplo).

In [229]:
from math import factorial #esta linea de importacion fue para sacar el factorial de una manera más sencilla
#Para este código se uso la clase calculadora, a la que se le agregaron funciones para hacer las respectivas operacione
class Calculadora: #se inicio con una clase calculadora
    def __init__(self, num1, num2): #asignamos los valores que tomará la clase para cada funcion
        self.num1 = float(num1) #incorporamos las variables como flotantes
        self.num2 = float(num2)

    def sumar(self): #iniciamos con las funciones, asignando las operaciones básicas (PARA CADA CASO) entre las variables antes definidas
        suma = self.num1 + self.num2
        print('El resultado de la suma de ', num1, 'mas ', num2, 'es: ', suma) #imprimos los resultados de una manera un tanto descriptiva

    def restar(self):
        resta = self.num1 - self.num2
        print('El resultado de la resta de ', num1, 'menos ', num2, 'es: ', resta)

    def multiplicar(self):
        multiplicacion = self.num1 * self.num2
        print('El resultado de la multiplicación de ', num1, 'por ', num2, 'es: ', multiplicacion)

    def dividir(self):
      try: #en este caso especial se asignaron excepciones a nuestra funcion como lo vimos en clase, para evitar tener un error al dividir entre cero, y así solo tener una advertencia
        division = self.num1 / self.num2
        print('El resultado de la divición de ', num1, 'sobre ', num2, 'es: ', division)
      except ZeroDivisionError as e:
        print('El segundo número (divisor) NO puede ser cero: ', e)

    def potencia(self):
        elevar = self.num1 ** self.num2
        print('El resultado de ', num1,'a la potencia ', num2,'es: ', int(elevar))#se hace una pequeña excepcion con un 'int' para obtener solo el numero entero de la potencia
        #Las siguintes líneas (31-38) se dejan como muestra de un intento fallido de incorporar una funcion para obtener el factorial, por separado funciona
"""    def factorial(n):
      if n==0 or n==1:
            resultado=1
      elif n > 1:
            resultado=n * producto(n-1)
      return resultado
      print('El factorial de ', n,'es: ', resultado)
"""
num1 = input('ingrese el primer numero: ') #estas lineas nos permiten manipular con que valores se haran las operaciones
num2 = input('ingrese el segundo numero: ')
n = int(input('ingrese un número entero positivo para obtener su factorial: ')) #esta variable se asigno para convertir el valor de 'num1' a entero y así se pueda sacar el factorial
print ('El factorial de ', n, 'es', factorial(n)) #con esta línea escribimos el factorial del primero numero introducido

  #incluimos las paqueterias correspondientes a cada operacion, menos la de factorial ya que no se pudo incorporar
calculadora = Calculadora(num1, num2)
calculadora.sumar()
calculadora.restar()
calculadora.multiplicar()
calculadora.dividir()
calculadora.potencia()
#calculadora.factorial()

ingrese el primer numero: 12.2
ingrese el segundo numero: 0
ingrese un número entero positivo para obtener su factorial: 46
El factorial de  46 es 5502622159812088949850305428800254892961651752960000000000
El resultado de la suma de  12.2 mas  0 es:  12.2
El resultado de la resta de  12.2 menos  0 es:  12.2
El resultado de la multiplicación de  12.2 por  0 es:  0.0
El segundo número (divisor) NO puede ser cero:  float division by zero
El resultado de  12.2 a la potencia  0 es:  1


Como podemos ver en el codigo de la calculadora, se trabajaron los errores
de division entre cero, sin embargo, al intentar meter fracciones no se lográ una ejecucion exitosa, así como al querer meter decimales, esto re "arreglo" pidiendole al usuario que para el caso del factorial solo meta números,así como se logro obtener solo el valor entero de una potencia con decimales.
 
La últimma parte queda pendiente...

**Ejercicio de puntos extras opcional (1 punto):**

Supóngamos que tenemos una lista de números enteros y queremos encontrar el número más grande. Por ejemplo, si la lista es [1, 2, 3, 4, 5] el número más grande es 5. Crea una función que encuentre el número más grande en una lista de números. La función debe regresar el número más grande y el índice en el que se encuentra. Prueba la función con una lista de números de tu elección.

In [ ]:
#Primero creamos una lista con los números a evaluar
lista = [0,1,2,3,4,5,1000,9646,48,6389,968,6956,566,306,98]

def mayor(listado): #creamos una función que nos brinde el número mayor de la lista
	alto=0 #esta linea es para tener una referencia y comenzar a evaluar la función
	for i in range(0,len(lista)): #este ciclo nos sirve para visitar cada uno de los elementos de la lista, empieza desde la posicion 0 hasta la longitud de toda nuestra lista
		if lista[i]>alto: #si el elemento actual que se esta visitando de la lista es mayor que la  variable definida antes
			alto=lista[i] #la variable será igual a ese elemento que es más grande
	return(alto) #ahora solo retornamos 
print('El número más grande de la lista ', lista, 'es: ' , mayor(lista)) #aqui solo nos queda imprimir la lista de la que se tomo el valor y el valor de nuestra funcion
print('La posicion del número más grande es: ', lista.index(mayor(lista))) #con la linea index obtenemos la posicion del número sacado de nuestra funcion

El número más grande de la lista  [0, 1, 2, 3, 4, 5, 1000, 9646, 48, 6389, 968, 6956, 566, 306, 98] es:  9646
La posicion del número más grande es:  7
